### PR TITLE
Introduce namespace resolver for validating and resolving stream names

### PR DESCRIPTION
**Pull request by jderrickk**

### CODE REVIEW

**Comment by sijie** [comment]
@jderrickk I think the change is making the namespace management become much complicated.

I had a brief discussion with @leighst offline about the naming management in DL. I think we should probably just stick into one naming convention. for example, we probably can just support zookeeper file-path like name. so 1) we can simply map the stream name to the path of zookeeper metadata; 2) this allow application manages how their streams will be organized in their own hierarchical.

**Comment by jderrickk** [comment]
@sijie sounds good to me. I am fine with single naming convention. I will update the pull request.

File: distributedlog-core/src/main/java/com/twitter/distributedlog/BKDistributedLogNamespace.java [hunk 1]
```diff
@@ -470,6 +478,23 @@ private BKDistributedLogNamespace(
...
+     *
+     * @param nameOfStream
+     *          name of stream
+     * @throws InvalidStreamNameException
+     */
+    private void validateName(String nameOfStream)
+            throws InvalidStreamNameException {
+        // validate the stream name
+        nsResolver.validateStreamName(nameOfStream);
+        if (isReservedStreamName(nameOfStream)) {
```
Comment (sijie) [inline distributedlog-core/src/main/java/com/twitter/distributedlog/BKDistributedLogNamespace.java]: shall we move the 'isReservedStreamName' validation to the resolver?

or shall we just deprecate using '.' for any component name in the path?

File: distributedlog-core/src/main/java/com/twitter/distributedlog/metadata/BKDLConfig.java [hunk 2]
```diff
@@ -19,9 +19,11 @@
 
 import com.google.common.annotations.VisibleForTesting;
 import com.google.common.base.Objects;
+import com.google.common.base.Optional;
```
Comment (sijie) [inline distributedlog-core/src/main/java/com/twitter/distributedlog/metadata/BKDLConfig.java]: unnecessary changes?

File: distributedlog-core/src/main/java/com/twitter/distributedlog/metadata/BKDLConfig.java [hunk 3]
```diff
@@ -19,9 +19,11 @@
...
 import com.twitter.distributedlog.DistributedLogConfiguration;
 import com.twitter.distributedlog.DistributedLogConstants;
 import com.twitter.distributedlog.ZooKeeperClient;
+import com.twitter.distributedlog.namespace.resolver.NamespaceResolver;
```
Comment (sijie) [inline distributedlog-core/src/main/java/com/twitter/distributedlog/metadata/BKDLConfig.java]: unnecessary changes?

**Comment by sijie** [comment]
@jderrickk are you still willing to contribute this?

I am thinking it might be just easier to accept zookeeper-like path as the stream name. we don't need to introduce namespace resolver.

**Comment by sijie** [comment]
@jderrickk ping?

**Comment by sijie** [comment]
@jderrickk #130 supports filesystem-like path. so I think we probably don't this any more, right?

**Comment by sijie** [comment]
Closed this pull request since #130 already did similar thing. And this pull request has been idle for a  long time.
